# Time analysis 
How do we make sense of the benchmarks generated by relevance and inference steps in the benchmarks.ipynb? This notebooks combines the benchmarks and gives conclusive times for pdfs. We used an a large memory intensive instance with 4 CPUs, 32Gb memory, and 1 GPU for our measurements. 

In [1]:
import pandas as pd
metrics_df = pd.read_pickle('../../reports/benchmarks/metrics_df.pkl')
kpi_metrics_df = pd.read_pickle('../../reports/benchmarks/kpi_metrics_df.pkl')

In [2]:
metrics_df.rename(columns={
    'PDF Name': 'pdf_name',
    'Number of Pages': 'pages',
    'Number of Data Points': 'rel_data_points',
    'Total Inference Time': 'rel_inference_time',
    'Time per data point': 'rel_time_per_point',
    'Data points per sec': 'rel_points_per_sec'}, inplace=True)
metrics_df.set_index('pdf_name', inplace=True)
metrics_df

pages  rel_data_points  \
pdf_name                                                                     
04_NOVATEK_AR_2016_ENG_11                             119            24264   
04_NOVATEK_AR_2018_ENG_15                             105            23112   
2013_book_mol_ar_eng_fin                              135            63024   
2015_BASF_Report                                      261            76392   
2017 Sustainability Report                             57            21936   
...                                                   ...              ...   
sustainability 2015                                   141            43920   
sustainability-report-2019                             32            17424   
sustainability-report-repsol-2016-eng-april-baj...     13             8088   
sustainable development 2017                          110            54240   
vopak_annual_report_2019                              224           132528   

                                                    rel_inference_time  \
pdf_name                                                                 
04_NOVATEK_AR_2016_ENG_11                                   169.834546   
04_NOVATEK_AR_2018_ENG_15                                   162.614384   
2013_book_mol_ar_eng_fin                                    447.409486   
2015_BASF_Report                                            530.416069   
2017 Sustainability Report                                  147.753352   
...                                                                ...   
sustainability 2015                                         292.777661   
sustainability-report-2019                                  118.627661   
sustainability-report-repsol-2016-eng-april-baj...           55.069756   
sustainable development 2017                                358.022435   
vopak_annual_report_2019                                    870.370404   

                                                    rel_time_per_point  \
pdf_name                                                                 
04_NOVATEK_AR_2016_ENG_11                                     0.006999   
04_NOVATEK_AR_2018_ENG_15                                     0.007036   
2013_book_mol_ar_eng_fin                                      0.007099   
2015_BASF_Report                                              0.006943   
2017 Sustainability Report                                    0.006736   
...                                                                ...   
sustainability 2015                                           0.006666   
sustainability-report-2019                                    0.006808   
sustainability-report-repsol-2016-eng-april-baj...            0.006809   
sustainable development 2017                                  0.006601   
vopak_annual_report_2019                                      0.006567   

                                                    rel_points_per_sec  
pdf_name                                                                
04_NOVATEK_AR_2016_ENG_11                                   142.868460  
04_NOVATEK_AR_2018_ENG_15                                   142.127648  
2013_book_mol_ar_eng_fin                                    140.864246  
2015_BASF_Report                                            144.022786  
2017 Sustainability Report                                  148.463636  
...                                                                ...  
sustainability 2015                                         150.011445  
sustainability-report-2019                                  146.879740  
sustainability-report-repsol-2016-eng-april-baj...          146.868273  
sustainable development 2017                                151.498886  
vopak_annual_report_2019                                    152.266207  

[144 rows x 5 columns]

In [3]:
kpi_metrics_df.rename(columns={
    'PDF Name': 'pdf_name',
    'Number of Data Points': 'kpi_data_points',
    'Total Inference Time': 'kpi_inference_time',
    'Time per data point': 'kpi_time_per_point',
    'Data points per sec': 'kpi_points_per_sec'}, inplace=True)
kpi_metrics_df.set_index('pdf_name', inplace=True)
kpi_metrics_df

kpi_data_points  \
pdf_name                                                              
China Resources Power Holdings Co Ltd Sustainab...             3619   
Uniper_Sustainability_Report_2019_EN                           3797   
Galp_Integrated_Report_2018                                   22454   
2018_sustainability_report                                     4166   
Annual-report-2019                                            14704   
...                                                             ...   
Enel SA sustainability-report-2018                            14917   
Sustainability Report 2017_EN                                  3557   
Eskom Holdings SOC Ltd Integrated Report 2019                 13347   
2018 Annual Report                                            10091   
sustainability-report-repsol-2016-eng-april-baj...              995   

                                                    kpi_inference_time  \
pdf_name                                                                 
China Resources Power Holdings Co Ltd Sustainab...           79.930830   
Uniper_Sustainability_Report_2019_EN                         83.328373   
Galp_Integrated_Report_2018                                 475.892238   
2018_sustainability_report                                   94.522472   
Annual-report-2019                                          405.279577   
...                                                                ...   
Enel SA sustainability-report-2018                          302.939695   
Sustainability Report 2017_EN                                77.206329   
Eskom Holdings SOC Ltd Integrated Report 2019               290.541965   
2018 Annual Report                                          223.097630   
sustainability-report-repsol-2016-eng-april-baj...           21.043432   

                                                    kpi_time_per_point  \
pdf_name                                                                 
China Resources Power Holdings Co Ltd Sustainab...            0.022086   
Uniper_Sustainability_Report_2019_EN                          0.021946   
Galp_Integrated_Report_2018                                   0.021194   
2018_sustainability_report                                    0.022689   
Annual-report-2019                                            0.027563   
...                                                                ...   
Enel SA sustainability-report-2018                            0.020308   
Sustainability Report 2017_EN                                 0.021705   
Eskom Holdings SOC Ltd Integrated Report 2019                 0.021768   
2018 Annual Report                                            0.022109   
sustainability-report-repsol-2016-eng-april-baj...            0.021149   

                                                    kpi_points_per_sec  
pdf_name                                                                
China Resources Power Holdings Co Ltd Sustainab...           45.276647  
Uniper_Sustainability_Report_2019_EN                         45.566712  
Galp_Integrated_Report_2018                                  47.182951  
2018_sustainability_report                                   44.074175  
Annual-report-2019                                           36.281127  
...                                                                ...  
Enel SA sustainability-report-2018                           49.240823  
Sustainability Report 2017_EN                                46.071353  
Eskom Holdings SOC Ltd Integrated Report 2019                45.938286  
2018 Annual Report                                           45.231319  
sustainability-report-repsol-2016-eng-april-baj...           47.283161  

[144 rows x 4 columns]

In [4]:
benchmark_df = metrics_df.join(kpi_metrics_df)

In [5]:
benchmark_df

pages  rel_data_points  \
pdf_name                                                                     
04_NOVATEK_AR_2016_ENG_11                             119            24264   
04_NOVATEK_AR_2018_ENG_15                             105            23112   
2013_book_mol_ar_eng_fin                              135            63024   
2015_BASF_Report                                      261            76392   
2017 Sustainability Report                             57            21936   
...                                                   ...              ...   
sustainability 2015                                   141            43920   
sustainability-report-2019                             32            17424   
sustainability-report-repsol-2016-eng-april-baj...     13             8088   
sustainable development 2017                          110            54240   
vopak_annual_report_2019                              224           132528   

                                                    rel_inference_time  \
pdf_name                                                                 
04_NOVATEK_AR_2016_ENG_11                                   169.834546   
04_NOVATEK_AR_2018_ENG_15                                   162.614384   
2013_book_mol_ar_eng_fin                                    447.409486   
2015_BASF_Report                                            530.416069   
2017 Sustainability Report                                  147.753352   
...                                                                ...   
sustainability 2015                                         292.777661   
sustainability-report-2019                                  118.627661   
sustainability-report-repsol-2016-eng-april-baj...           55.069756   
sustainable development 2017                                358.022435   
vopak_annual_report_2019                                    870.370404   

                                                    rel_time_per_point  \
pdf_name                                                                 
04_NOVATEK_AR_2016_ENG_11                                     0.006999   
04_NOVATEK_AR_2018_ENG_15                                     0.007036   
2013_book_mol_ar_eng_fin                                      0.007099   
2015_BASF_Report                                              0.006943   
2017 Sustainability Report                                    0.006736   
...                                                                ...   
sustainability 2015                                           0.006666   
sustainability-report-2019                                    0.006808   
sustainability-report-repsol-2016-eng-april-baj...            0.006809   
sustainable development 2017                                  0.006601   
vopak_annual_report_2019                                      0.006567   

                                                    rel_points_per_sec  \
pdf_name                                                                 
04_NOVATEK_AR_2016_ENG_11                                   142.868460   
04_NOVATEK_AR_2018_ENG_15                                   142.127648   
2013_book_mol_ar_eng_fin                                    140.864246   
2015_BASF_Report                                            144.022786   
2017 Sustainability Report                                  148.463636   
...                                                                ...   
sustainability 2015                                         150.011445   
sustainability-report-2019                                  146.879740   
sustainability-report-repsol-2016-eng-april-baj...          146.868273   
sustainable development 2017                                151.498886   
vopak_annual_report_2019                                    152.266207   

                                                    kpi_data_points  \
pdf_name                                                              
04_NOV

In [6]:
# Clean dataframe with an entry with 0 pages
benchmark_df.loc['Maharashtra State Power Generation Co. Ltd._website', 'pages'] = 1

## Finding average number of pages in our pdfs

In [7]:
benchmark_df['pages'].describe()

count    144.000000
mean     156.965278
std      117.540733
min        1.000000
25%       73.250000
50%      127.500000
75%      224.250000
max      653.000000
Name: pages, dtype: float64

Each pdf has on average 157 pages, 75% of pdfs have pages less than 224. The median number of pages is 127 and max is 653.

## Finding average time for relevance

In [8]:
temp = benchmark_df['rel_data_points'] / benchmark_df['pages']
temp.describe()

count     144.000000
mean      359.037705
std       181.697295
min        73.066667
25%       239.955882
50%       305.102564
75%       434.983559
max      1042.036364
dtype: float64

In [9]:
benchmark_df['rel_time_per_point'].describe()

count    144.000000
mean       0.006918
std        0.000771
min        0.006433
25%        0.006697
50%        0.006809
75%        0.006922
max        0.013642
Name: rel_time_per_point, dtype: float64

The average time per data point is 0.006918s. Note the std is quite low compared to the order of mean, indicating that it is a reliable value across data points.

A pdf with 157 pages, and 360 data points per page, will take 391 seconds or 6.5min to execute

In [10]:
benchmark_df['rel_inference_time'].describe()

count     144.000000
mean      409.545597
std       448.180519
min         3.240382
25%       140.197153
50%       284.229863
75%       552.409658
max      3078.194455
Name: rel_inference_time, dtype: float64

The average time for a pdf overall is 409 seconds or 6.8 mins with a std of 7.5mins. Also, 75% of the pdfs will take between a few seconds to 9mins. The median pdfs will take less than 5mins.

## Finding average time for kpi

In [11]:
temp = benchmark_df['kpi_data_points'] / benchmark_df['pages']
temp.describe()

count    144.000000
mean      84.259906
std       43.973948
min       16.444444
25%       49.613165
50%       75.081418
75%      115.574053
max      238.000000
dtype: float64

In [12]:
benchmark_df['kpi_time_per_point'].describe()

count    144.000000
mean       0.022305
std        0.002467
min        0.020149
25%        0.021308
50%        0.021803
75%        0.022309
max        0.039455
Name: kpi_time_per_point, dtype: float64

The average time per data point is 0.022305s. Note the std is quite low compared to the order of mean, indicating that it is a reliable value across data points.

A pdf with 157 pages, and 85 data points per page, will take 298 seconds or ~5min to execute

In [13]:
benchmark_df['kpi_inference_time'].describe()

count     144.000000
mean      323.633887
std       339.327505
min         0.677586
25%        79.557247
50%       205.765385
75%       461.444348
max      1817.714101
Name: kpi_inference_time, dtype: float64

The average time for a pdf overall is 324 seconds or 5.4 mins with a std of 5.6mins. Also, 75% of the pdfs will take between a few seconds to 8mins. The median pdfs will take less than 3.4mins.

Total time for both the steps

We can add the averages for each step. For a pdf with 157 pages, it will take around 11.5mins to execute. We can also look at the total time and then average it.

In [14]:
benchmark_df['total_inference_time'] = benchmark_df['rel_inference_time'] + benchmark_df['kpi_inference_time']

In [15]:
benchmark_df['total_inference_time'].describe()

count     144.000000
mean      733.179484
std       758.515997
min         3.917968
25%       209.536250
50%       496.674452
75%      1046.226577
max      4727.612188
Name: total_inference_time, dtype: float64

This means average total time is 733sec or ~12min with a std of another 12mins. 75% of the pdfs will take a few secs to 17.4 mins. The maximum time it took for a pdf in our list was 78mins and the minimum was 4secs. 

## Total time it took for 144 pdfs

In [16]:
benchmark_df['total_inference_time'].sum()/(60*60)

29.327179345223637

### 29.3hrs was the total time taken for 144 pdfs.

# Conclusion

An average pdf with around 150 pages will take 12 mins for the relevance and kpi models to infer. The pdfs in our datasets range from a single page to a 634 pages long and they take 4sec to 78mins for inference. 